File: mini-claw-code-py/src/mini_claw_code_py/tools/bash.py
from __future__ import annotations

import asyncio
from typing import Any

from ..types import ToolDefinition


class BashTool:
    def __init__(self) -> None:
        self._definition = ToolDefinition.new(
            "bash",
            "Run a bash command and return its output.",
        ).param("command", "string", "The bash command to run", True)

    @property
    def definition(self) -> ToolDefinition:
        return self._definition

    async def call(self, args: Any) -> str:
        command = args.get("command") if isinstance(args, dict) else None
        if not isinstance(command, str):
            raise ValueError("missing 'command' argument")

        process = await asyncio.create_subprocess_exec(
            "bash",
            "-lc",
            command,
            stdout=asyncio.subprocess.PIPE,
            stderr=asyncio.subprocess.PIPE,
        )
        stdout_bytes, stderr_bytes = await process.communicate()
        stdout = stdout_bytes.decode()
        stderr = stderr_bytes.decode()

        parts: list[str] = []
        if stdout:
            parts.append(stdout)
        if stderr:
            parts.append(f"stderr: {stderr}")
        if not parts:
            return "(no output)"
        return "\n".join(parts)

File: mini-claw-code-ts/src/tools/bash.ts
import { $ } from "bun";
import { ToolDefinition, type JsonObject, type Tool } from "../types";

export class BashTool implements Tool {
  readonly #definition = new ToolDefinition(
    "bash",
    "Run a bash command and return its output.",
  ).param("command", "string", "The bash command to run", true);

  definition(): ToolDefinition {
    return this.#definition;
  }

  async call(args: JsonObject | null): Promise<string> {
    const command = args?.command;
    if (typeof command !== "string") {
      throw new Error("missing 'command' argument");
    }

    const proc = Bun.spawn({
      cmd: ["bash", "-lc", command],
      stdout: "pipe",
      stderr: "pipe",
    });

    const stdout = await new Response(proc.stdout).text();
    const stderr = await new Response(proc.stderr).text();
    await proc.exited;

    const parts: string[] = [];
    if (stdout.trimEnd().length > 0) {
      parts.push(stdout.trimEnd());
    }
    if (stderr.trimEnd().length > 0) {
      parts.push(`stderr: ${stderr.trimEnd()}`);
    }

    return parts.length > 0 ? parts.join("\n") : "(no output)";
  }
}
